cdc/sink/mq.go

```diff
@@ -164,19 +177,59 @@ func (k *mqSink) EmitRowChangedEvents(ctx context.Context, rows ...*model.RowCha
 	return nil
 }
 
+// FlushRowChangedEvents is thread-safety
 func (k *mqSink) FlushRowChangedEvents(ctx context.Context, tableID model.TableID, resolvedTs uint64) (uint64, error) {
-	if checkpointTs, ok := k.tableCheckpointTs[tableID]; ok && resolvedTs <= checkpointTs {
+	var checkpointTs uint64
+	v, ok := k.tableCheckpointTsMap.Load(tableID)
+	if ok {
+		checkpointTs = v.(uint64)
+	}
+	if resolvedTs <= checkpointTs {
 		return checkpointTs, nil
 	}
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case k.resolvedBuffer <- resolvedTsEvent{
+		tableID:    tableID,
+		resolvedTs: resolvedTs,
+	}:
+	}
+	k.statistics.PrintStatus(ctx)
+	return checkpointTs, nil
+}
+
+// bgFlushTs flush resolvedTs to workers and flush the mqProducer
+func (k *mqSink) bgFlushTs(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case msg := <-k.resolvedBuffer:
+			resolvedTs := msg.resolvedTs
+			err := k.flushTsToWorker(ctx, resolvedTs)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			err = k.mqProducer.Flush(ctx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			// Since CDC does not guarantee exactly once semantic, it won't cause any problem
+			// here even if the table was moved or removed.
+			// ref: https://github.com/pingcap/tiflow/pull/4356#:~:text=What%20if%20a%20table%20is%20removed%20between%20line%20213%20and%20214%3F
```

Comment:
```suggestion
			// ref: https://github.com/pingcap/tiflow/pull/4356#discussion_r787405134
```

Let's shorten the link.